src/engine/strat_engine/thinpool.rs

```diff
@@ -73,74 +91,100 @@ impl ThinPool {
+                                                   fssave.thin_id,
+                                                   fssave.size));
+                Ok(StratFilesystem::setup(fssave.uuid, &fssave.name, thin_dev))
+            };
+
+            try!(fs_save.iter().map(get_filesystem).collect())
+        };
+
+        let mut fs_table = Table::default();
+        for fs in filesystems {
+            let evicted = fs_table.insert(fs);
+            if !evicted.is_empty() {
+                let err_msg = "filesystems with duplicate UUID or name specified in metadata";
+                return Err(EngineError::Engine(ErrorEnum::Invalid, err_msg.into()));
             }
-            Err(err) => Err(err.into()),
         }
+
+        Ok(ThinPool {
+               thin_pool: thinpool_dev,
+               meta_spare: spare_segments,
+               id_gen: ThinDevIdPool::new_from_ids(thin_ids),
+               filesystems: fs_table,
+               mdv: mdv,
+           })
     }
 
     /// The status of the thin pool as calculated by DM.
     pub fn thin_pool_status(&self, dm: &DM) -> EngineResult<ThinPoolStatus> {
-        Ok(try!(self.thin_pool.status(dm)))
-    }
+        let thinpool = try!(self.thin_pool.status(dm));
+        try!(self.mdv.check());
 
-    /// Make a new thin device.
-    pub fn make_thin_device(&mut self,
-                            dm: &DM,
-                            name: &str,
-                            size: Option<Sectors>)
-                            -> EngineResult<ThinDev> {
-        Ok(try!(ThinDev::new(name,
-                             dm,
-                             &self.thin_pool,
-                             try!(self.id_gen.new_id()),
-                             size.unwrap_or(DEFAULT_THIN_DEV_SIZE))))
-    }
+        let filesystems: Vec<_> = try!(self.filesystems
+                                           .into_iter()
+                                           .map(|fs| fs.check(dm))
+                                           .collect());
 
-    /// Setup a previously constructed thin device.
-    pub fn setup_thin_device(&self,
-                             dm: &DM,
-                             name: &str,
-                             id: ThinDevId,
-                             size: Sectors)
-                             -> EngineResult<ThinDev> {
-        Ok(try!(ThinDev::setup(name, dm, &self.thin_pool, id, size)))
+        Ok(ThinPoolStatus {
+               thinpool,
+               filesystems,
+           })
     }
 
-    /// Tear down the thin pool.
-    pub fn teardown(self, dm: &DM) -> DmResult<()> {
-        self.thin_pool.teardown(dm)
+    /// Tear down the components managed here: filesystems, the MDV,
+    /// and the actual thinpool device itself.
+    pub fn teardown(self, dm: &DM) -> EngineResult<()> {
+        for fs in self.filesystems.empty() {
+            try!(fs.teardown(dm));
+        }
+
+        try!(self.thin_pool.teardown(dm));
```

Comment:
I think this is best addressed by explicit comments in this method on which teardowns depend on other things succeeding, and which don't, so that's the approach I've taken.